Content.Server/Mind/MindSystem.cs

```diff
@@ -502,87 +527,126 @@ public bool TryRemoveObjective(Mind mind, int index)
+    /// <typeparam name="T">The role type to add.</typeparam>
     /// <returns>The instance of the role.</returns>
     /// <exception cref="ArgumentException">
     ///     Thrown if we already have a role with this type.
     /// </exception>
-    public void AddRole(Mind mind, Role role)
+    public void AddRole<T>(EntityUid mindId, T component, MindComponent? mind = null) where T : Component, new()
     {
-        if (mind.Roles.Contains(role))
+        if (!Resolve(mindId, ref mind))
+            return;
+
+        if (HasComp<T>(mindId))
         {
-            throw new ArgumentException($"We already have this role: {role}");
+            throw new ArgumentException($"We already have this role: {typeof(T)}");
         }
 
-        mind.Roles.Add(role);
-        role.Greet();
+        AddComp(mindId, component);
+        var antagonist = _roles.IsAntagonistRole<T>();
 
-        var message = new RoleAddedEvent(mind, role);
+        var message = new RoleAddedEvent(mindId, mind, antagonist);
         if (mind.OwnedEntity != null)
         {
             RaiseLocalEvent(mind.OwnedEntity.Value, message, true);
         }
 
         _adminLogger.Add(LogType.Mind, LogImpact.Low,
-            $"'{role.Name}' added to mind of {MindOwnerLoggingString(mind)}");
+            $"'Role {typeof(T).Name}' added to mind of {MindOwnerLoggingString(mind)}");
+
+        return;
     }
 
     /// <summary>
     ///     Removes a role from this mind.
     /// </summary>
-    /// <param name="mind">The mind to remove the role from.</param>
-    /// <param name="role">The type of the role to remove.</param>
+    /// <param name="mindId">The mind to remove the role from.</param>
+    /// <typeparam name="T">The type of the role to remove.</typeparam>
     /// <exception cref="ArgumentException">
     ///     Thrown if we do not have this role.
     /// </exception>
-    public void RemoveRole(Mind mind, Role role)
+    public void RemoveRole<T>(EntityUid mindId) where T : Component
     {
-        if (!mind.Roles.Contains(role))
+        if (!RemComp<T>(mindId))
         {
-            throw new ArgumentException($"We do not have this role: {role}");
+            throw new ArgumentException($"We do not have this role: {typeof(T)}");
         }
 
-        mind.Roles.Remove(role);
-
-        var message = new RoleRemovedEvent(mind, role);
+        var mind = Comp<MindComponent>(mindId);
+        var antagonist = _roles.IsAntagonistRole<T>();
+        var message = new RoleRemovedEvent(mindId, mind, antagonist);
 
         if (mind.OwnedEntity != null)
         {
             RaiseLocalEvent(mind.OwnedEntity.Value, message, true);
         }
         _adminLogger.Add(LogType.Mind, LogImpact.Low,
-            $"'{role.Name}' removed from mind of {MindOwnerLoggingString(mind)}");
+            $"'Role {typeof(T).Name}' removed from mind of {MindOwnerLoggingString(mind)}");
+    }
+
+    public void TryRemoveRole<T>(EntityUid mindId) where T : Component
```

Comment:
bool or rename?

Reply:
I originally called it Try as it doesn't throw when the role doesn't exist, unlike RemoveRole, but I changed it now in [`e533e6e` (#19591)](https://github.com/space-wizards/space-station-14/pull/19591/commits/e533e6ecbc06a601745cb2a8b865c289615f1759)
Should RemoveRole just not throw at all, like RemoveComponent?